groups/reconcile.go

```diff
@@ -181,69 +184,176 @@ func main() {
 	client := credential.Client(ctx)
 	clientOption := option.WithHTTPClient(client)
 
-	adminService, err := admin.NewService(ctx, clientOption)
+	r, err := newReconciler(ctx, clientOption)
 	if err != nil {
-		log.Fatalf("Unable to retrieve directory Client %v", err)
-	}
-
-	groupsSettingsService, err := groupssettings.NewService(ctx, clientOption)
-	if err != nil {
-		log.Fatalf("Unable to retrieve groupssettings Service %v", err)
+		log.Fatal(err)
 	}
 
 	if *printConfig {
-		err = printGroupMembersAndSettings(adminService, groupsSettingsService)
+		err = r.printGroupMembersAndSettings()
 		if err != nil {
 			log.Fatal(err)
 		}
 		return
 	}
 
-	log.Println(" ======================= Updates =========================")
-	for _, g := range groupsConfig.Groups {
+	log.Println(" ======================= Updates =======================")
+	err = r.reconcileGroups(groupsConfig.Groups)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+// reconciler syncs the actual state of the world with the configuration.
+// It does so by making use of AdminService and GroupService which are mockable
+// interfaces.
+type reconciler struct {
+	adminService AdminService
+	groupService GroupService
+}
+
+func newReconciler(ctx context.Context, clientOption option.ClientOption) (*reconciler, error) {
+	as, err := NewAdminService(ctx, clientOption)
+	if err != nil {
+		return nil, err
+	}
+
+	gs, err := NewGroupService(ctx, clientOption)
+	if err != nil {
+		return nil, err
+	}
+
+	return &reconciler{adminService: as, groupService: gs}, nil
+}
+
+func (r *reconciler) reconcileGroups(groups []GoogleGroup) error {
+	for _, g := range groups {
 		if g.EmailId == "" {
-			log.Fatal(fmt.Sprintf("Group has no email-id: %#v", g))
+			return fmt.Errorf("group has no email-id: %#v", g)
 		}
 
-		err = createOrUpdateGroupIfNecessary(adminService, g.EmailId, g.Name, g.Description)
+		// update the group that is currently being considered.
+		r.adminService.SetGroup(g)
+		r.groupService.SetGroup(g)
```

Comment:
This context-aware API seems prone to human error.  I would rather either:
- make group an explicit parameter to function calls (e.g. `adminService.CreateOrUpdateGroupIfNescessary(g)`)
- return and use an instance of something that is configured to use group (e.g. `a = adminServiceFactory.AdminServiceForGroup(g)` but those are terrible names I'm sure you can think of something better)